dagstore_gc.go

```diff
@@ -73,6 +74,80 @@ func (d *DAGStore) gc(resCh chan *GCResult) {
 	}
 }
 
+// gc performs DAGStore GC. Refer to DAGStore#GC for more information.
+//
+// The event loops gives it exclusive execution rights, so while GC is running,
+// no other events are being processed.
+func (d *DAGStore) lruGC(target float64) {
+	// sort shards by LRU
+	d.lk.RLock()
+	var reclaim []*Shard
+	for _, s := range d.shards {
+		s.lk.RLock()
+		reclaim = append(reclaim, s)
+		s.lk.RUnlock()
+	}
+	d.lk.RUnlock()
+
+	sort.Slice(reclaim, func(i, j int) bool {
+		reclaim[i].lk.RLock()
+		defer reclaim[i].lk.RUnlock()
+
+		reclaim[j].lk.RLock()
+		defer reclaim[j].lk.RUnlock()
+
+		return reclaim[i].lastAccessedAt.Before(reclaim[j].lastAccessedAt)
+	})
+
+	// TODO What should we do about shards that are currently being initialised/acquired ?
```

Comment:
This leaks. This should be part of a LRUGarbageCollector.

Reply:
By storing the access metadata in the garbage collector, you can likely remove much of the locking.

Reply:
Potential interfaces.

If garbage collector manages the directory directly:

```go
type GarbageCollector interface {
    NotifyAccessed(*shard.Shard) // invoked every time an acquirer requests the shard
    NotifyReclaimable(*shard.Shard) // when shard goes back from Serving to Available; we could notify every time we release, but then the GarbageCollector would need to do usage tracking, which would be a duplication of responsibliities
    NotifyRemoved(*shard.Shard) // stop tracking this shard, e.g. when a shard is terminally destroyed and the dagstore has already removed its resources
    Reclaim() []*shard.Shard // runs GC and returns which shards were reclaimed
}
```

If garbage collector only advises which shards are reclaimable, but does not own the management of the transients dir (the dagstore does):

```go
type GarbageCollector interface {
    NotifyAccessed(shard.Key) // invoked every time an acquirer requests the shard
    NotifyReclaimable(shard.Key) // when shard goes back from Serving to Available; we could notify every time we release, but then the GarbageCollector would need to do usage tracking, which would be a duplication of responsibliities
    NotifyRemoved(shard.Key) // stop tracking this shard, e.g. when a shard is terminally destroyed and the dagstore has already removed its resources

    Reclaimable() []shard.Key // returns all shards ordered by reclaimability precedence
    NotifyReclaimed([]shard.Key) // informs which shards were ultimately reclaimed
}
```